internal/pkg/api/handleCheckin.go

```diff
@@ -917,50 +922,50 @@ func parseMeta(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]
 	return outMeta, nil
 }
 
-func parseComponents(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]byte, error) {
+func parseComponents(zlog zerolog.Logger, agent *model.Agent, req *CheckinRequest) ([]byte, *[]string, error) {
+	var unhealthyReason []string
+
+	// fallback to other if components don't exist
+	if agent.UnhealthyReason == nil && (agent.LastCheckinStatus == FailedStatus || agent.LastCheckinStatus == DegradedStatus) {
+		unhealthyReason = []string{"other"}
+	}
+
 	if req.Components == nil {
-		return nil, nil
+		return nil, &unhealthyReason, nil
+	}
+
+	agentComponentsJSON, err := json.Marshal(agent.Components)
+	if err != nil {
+		return nil, &unhealthyReason, fmt.Errorf("agent.Components marshal: %w", err)
 	}
 
 	// Quick comparison first; compare the JSON payloads.
 	// If the data is not consistently normalized, this short-circuit will not work.
-	if bytes.Equal(*req.Components, agent.Components) {
+	if bytes.Equal(*req.Components, agentComponentsJSON) {
 		zlog.Trace().Msg("quick comparing agent components data is equal")
-		return nil, nil
+		return nil, &unhealthyReason, nil
 	}
 
 	// Deserialize the request components data
-	var reqComponents interface{}
+	var reqComponents []model.ComponentsItems
 	if len(*req.Components) > 0 {
 		if err := json.Unmarshal(*req.Components, &reqComponents); err != nil {
-			return nil, fmt.Errorf("parseComponents request: %w", err)
-		}
-		// Validate that components is an array
-		if _, ok := reqComponents.([]interface{}); !ok {
-			return nil, errors.New("parseComponets request: components property is not array")
+			return nil, &unhealthyReason, fmt.Errorf("parseComponents request: %w", err)
 		}
 	}
 
 	// If empty, don't step on existing data
 	if reqComponents == nil {
-		return nil, nil
-	}
-
-	// Deserialize the agent's components copy
-	var agentComponents interface{}
-	if len(agent.Components) > 0 {
-		if err := json.Unmarshal(agent.Components, &agentComponents); err != nil {
-			return nil, fmt.Errorf("parseComponents local: %w", err)
-		}
+		return nil, &unhealthyReason, nil
 	}
 
 	var outComponents []byte
 
 	// Compare the deserialized meta structures and return the bytes to update if different
-	if !reflect.DeepEqual(reqComponents, agentComponents) {
+	if !reflect.DeepEqual(reqComponents, agent.Components) {
```

Comment:
Do we have already a test that test we do not update if not different?

Reply:
added tests to verify when components equals / not equals